### PR TITLE
pkgs/top-level/config.nix: nixfmt, declare options

### LINKED ISSUE #330627: warnUndeclaredOptions warns about packageOverrides
### Steps To Reproduce
1. Add `nixpkgs.config.warnUndeclaredOptions = true;` to nixos config;
2. Add `nixpkgs.config.packageOverrides = pkgs: {};` to nixos config;
3. `nixos-rebuild build`

Output: `trace: warning: undeclared Nixpkgs option set: config.packageOverrides`

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.32-hardened1, NixOS, 24.05 (Uakari), 24.05.3164.63d37ccd2d17`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - channels(nix): `""`
 - channels(root): `"home-manager-24.05.tar.gz, nixos-24.05, nixos-hardware, nixos-unstable, nur"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by eclairevoyant**
## Description of changes

Reformat with `nixfmt-rfc-style`, and declare some options:
- `allowNonSource`
- `allowUnfreePredicate`
- `allowNonSourcePredicate`
- `packageOverrides`
  - fixes #330627
- `permittedInsecurePackages`
- `allowInsecurePredicate`
- `checkMetaRecursively`
- `handleEvalIssue`
- `inHydra`
- `allowlistedLicenses`, `blocklistedLicenses`

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/top-level/config.nix [hunk 1]
```diff
@@ -140,9 +182,26 @@ let
       feature = "build packages with ROCm support by default";
     };
 
+    packageOverrides = mkOption {
+      type = types.functionTo types.raw;
+      default = lib.id;
```
Comment (eclairevoyant) [inline pkgs/top-level/config.nix:187]: This default is due to https://github.com/NixOS/nixpkgs/actions/runs/10137454574/job/28027709399?pr=330753

File: pkgs/top-level/config.nix [hunk 2]
```diff
@@ -140,9 +182,26 @@ let
       feature = "build packages with ROCm support by default";
     };
 
+    packageOverrides = mkOption {
```
Comment (eclairevoyant) [inline pkgs/top-level/config.nix:185]: We might also want to deprecate this.

File: pkgs/top-level/config.nix [hunk 3]
```diff
@@ -140,9 +182,26 @@ let
...
+      type = types.functionTo types.raw;
```
Comment (eclairevoyant) [inline pkgs/top-level/config.nix:186]: Should this be `functionTo lazyAttrsOf raw` instead of `functionTo raw`?

File: pkgs/top-level/config.nix [hunk 4]
```diff
@@ -104,6 +108,80 @@ let
...
+            fullName = "Abstyles License";
+          })
+        ]
+      '';
+      description = ''
+        Disallows evaluation of a package, if the package has a license in this blocklist.
+      '';
+    };
+
+    allowNonSource = mkEnableOption "" // {
```
Comment (philiptaron) [inline pkgs/top-level/config.nix:160]: I did look askance at this `mkEnableOption ""` business, but there are other examples of it throughout the tree.

File: pkgs/top-level/config.nix [hunk 5]
```diff
@@ -104,6 +108,80 @@ let
...
+      default = true;
+      defaultText = literalExpression ''true && builtins.getEnv "NIXPKGS_ALLOW_NONSOURCE" != "0"'';
```
Comment (philiptaron) [inline pkgs/top-level/config.nix:162]: This is a funny default 😁

**Review by philiptaron — APPROVED** [review]
I took a look at the updated `nixpkgs-manual` as well. Looks good.

**Comment by philiptaron** [comment]
I plan on merging this on Monday unless I hear from others about Issues.

File: pkgs/stdenv/generic/check-meta.nix [hunk 6]
```diff
@@ -55,7 +55,7 @@ let
     envVar = builtins.getEnv "NIXPKGS_ALLOW_NONSOURCE";
   in if envVar != ""
      then envVar != "0"
-     else config.allowNonSource or true;
+     else config.allowNonSource;
```
[the remark this answers is not in the document]
Comment (eclairevoyant) [reply pkgs/stdenv/generic/check-meta.nix:58]: yeah, there are many inconsistencies in both the API and the impl. I'd eventually like to converge them, but for now I wanted to preserve the behaviour as a first phase.

File: pkgs/stdenv/generic/check-meta.nix [hunk 7]
```diff
@@ -47,7 +47,7 @@ let
   # import <nixpkgs> { config = { showDerivationWarnings = [ "maintainerless" ]; }; }
   showWarnings = config.showDerivationWarnings;
 
-  getNameWithVersion = attrs: attrs.name or ("${attrs.pname or "«name-missing»"}-${attrs.version or "«version-missing»"}");
+  getNameWithVersion = attrs: attrs.name or "${attrs.pname or "«name-missing»"}-${attrs.version or "«version-missing»"}";
```
Comment (philiptaron) [inline pkgs/stdenv/generic/check-meta.nix:50]: Love that nixf-tidy made this happen. 🎉

**Review by philiptaron — APPROVED** [review]

**Review by philiptaron — APPROVED** [review]
Did a final `nixpkgs-manual` build, diff'd it with master, looks good 👍🏻

**Comment by spiage** [comment]
@infinisil Hi! this breaks master's build
https://github.com/NixOS/nixpkgs/commit/42d0012a8bf81c807d9a581b6dd52f56a09fbd70#commitcomment-145046128

**Comment by philiptaron** [comment]
@spiage could you provide a link to your system's flake.nix? I don't see an "a7" in https://github.com/spiage/nixos-config/blob/main/flake.nix

**Comment by spiage** [comment]
well, I was not ready to open this machine's config, but OK, give me 2 minutes

**Comment by eclairevoyant** [comment]
It's reproducible by adding an insecure package to a nixos config. Maybe we should revert for now?

**Comment by spiage** [comment]
I'm using yandex-browser
and it have this lines https://github.com/NixOS/nixpkgs/blob/aee09d704ac7288bba5251b382286f4523c1f00e/pkgs/applications/networking/browsers/yandex-browser/default.nix#L169

**Comment by philiptaron** [comment]
> It's reproducible by adding an insecure package to a nixos config. Maybe we should revert for now?

Yeah, unless the fix is super simple. FWIW, I have [an unfree package](https://github.com/philiptaron/flock.nix/blob/lex/programs.nix#L272-L274) in my flake and no issues.

**Comment by spiage** [comment]
if it is my local problem I can use any walkthrough if you give me =)

**Comment by eclairevoyant** [comment]
Basically I didn't notice that the `config.foo or ...` constructions would no longer work, and it would either have to be replaced with a check on `options.foo.isDefined` or by explicitly setting the `default`

**Comment by philiptaron** [comment]
> Basically I didn't notice that the `config.foo or ...` constructions would no longer work, and it would either have to be replaced with a check on `options.foo.isDefined` or by explicitly setting the `default`

That's pretty hefty search-and-replace, and definitely a breaking change for external users. Setting the `default` is the least impactful option, as I see it, especially if there's a common and obvious default. Do you agree?

**Comment by spiage** [comment]
> @spiage could you provide a link to your system's flake.nix? I don't see an "a7" in https://github.com/spiage/nixos-config/blob/main/flake.nix

I am very ashamed to show it, but here is the config
https://github.com/spiage/nixos-config-a7

**Comment by eclairevoyant** [comment]
> > Basically I didn't notice that the `config.foo or ...` constructions would no longer work, and it would either have to be replaced with a check on `options.foo.isDefined` or by explicitly setting the `default`
> 
> That's pretty hefty search-and-replace, and definitely a breaking change for external users. Setting the `default` is the least impactful option, as I see it, especially if there's a common and obvious default. Do you agree?

Yes, I think setting `default` would be safer, but I don't have enough time to test right now. I'm okay with reverting for now and making a new PR with some test cases added.

**Comment by spiage** [comment]
I'm ready to help with tests if you need it =)

**Comment by philiptaron** [comment]
Thanks for the super-fast report @spiage! I'm going to merge @eclairevoyant's revert and you'll be on your way.

**Comment by spiage** [comment]
I have 10 minutes lag to react =)
https://github.com/spiage/nixos-config-a7/blob/1a5eb091e24f253eb2a48c620e142e89704d66f3/nu.sh#L7

**Comment by spiage** [comment]
It's ok after revert
```
Version 15449 -> 15450:
  nixos-system-a7: 24.11.20240805.6be7a53 → 24.11.20240805.fb7d848
  source: +14.1 KiB
```

**Comment by philiptaron** [comment]
@eclairevoyant, I'd love to get this PR rolling again. What do you see as the set of steps to get it in without causing a hullabaloo?

**Comment by eclairevoyant** [comment]
free time to write some tests :weary:

**Pull request merged**